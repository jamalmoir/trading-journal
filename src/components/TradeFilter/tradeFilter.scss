.trade-filter-panel {
  margin-bottom: 2em;
  padding: 1em;

  .trade-filter-group {
    .trade-filter {
      margin-bottom: 0.5em;
    }
  }
}