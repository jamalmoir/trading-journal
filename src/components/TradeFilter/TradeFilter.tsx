import * as React from 'react'
import { useState } from 'react'
import { connect } from 'react-redux'
import Types from 'Types'
import { TextInput } from '../TextInput'
import { Dispatch } from 'redux'
import ReactDatePicker from 'react-datepicker'
import { TristateCheckbox } from '../TristateCheckbox'
import ReactTags, { Tag } from 'react-tag-autocomplete'
import './tradeFilter.scss'
import { setTradeFilters } from '../../redux/actions/trade'
import { SelectInput } from '../SelectInput'

interface TradeFilterProps {
  onSetTradeFilters: (filters: Types.TradeFilters) => null
}

interface Tags {
  [index: string]: any
  tags: {
    tags: Tag[]
    suggestions: Tag[]
  }
  emotions: {
    tags: Tag[]
    suggestions: Tag[]
  }
}

const kindChoices = [
  { id: 'long', name: 'Long' },
  { id: 'short', name: 'Short' },
]

const ratingChoices = [
  { id: -1, name: '-1' },
  { id: 0, name: '0' },
  { id: 1, name: '1' },
]

export const TradeFilterComponent = (props: TradeFilterProps) => {
  const initialFilters: Types.TradeFilters = {
    instrument: null,
    strategy: null,
    kind: null,
    rating: null,
    entryDate: null,
    exitDate: null,
    profit: null,
    hitTakeProfit: null,
    flagged: null,
    managed: null,
    tags: null,
    emotions: null,
  }

  const intialTags: Tags = {
    tags: {
      tags: [],
      suggestions: [],
    },
    emotions: {
      tags: [],
      suggestions: [],
    },
  }

  const [filters, setFilters]: [
    Types.TradeFilters,
    (val: any) => void
  ] = useState(initialFilters)
  const [tags, setTags]: [Tags, (val: any) => void] = useState(intialTags)

  let updateFilter = (
    name: string,
    value: string | string[] | number | boolean | Date | null
  ) => {
    if (value === '') {
      value = null
    }

    const newFilters = { ...filters, [name]: value }

    setFilters(newFilters)
    props.onSetTradeFilters(newFilters)
  }

  let handleTagDelete = (kind: string, i: number) => {
    let stateKey: string

    if (kind === 'tags') {
      stateKey = 'tags'
    } else if (kind === 'emotions') {
      stateKey = 'emotions'
    }

    const newTags = {
      ...tags,
      [stateKey]: {
        ...tags[stateKey],
        tags: tags[stateKey].tags.filter(
          (tag: Tag, index: number) => index !== i
        ),
      },
    }

    setTags(newTags)

    updateFilter(stateKey, newTags[stateKey].tags.map((t: Tag) => t.name))
  }

  let handleTagAddition = (kind: string, tag: Tag) => {
    let stateKey: string

    if (kind === 'tags') {
      stateKey = 'tags'
    } else if (kind === 'emotions') {
      stateKey = 'emotions'
    }

    const newTags = {
      ...tags,
      [stateKey]: {
        ...tags[stateKey],
        tags: [...tags[stateKey].tags, tag],
      },
    }

    setTags(newTags)

    updateFilter(stateKey, newTags[stateKey].tags.map((t: Tag) => t.name))
  }

  return (
    <div className="trade-filter-panel grid-x grid-margin-y">
      <div className="cell">Filters</div>
      <TextInput
        type="text"
        className="trade-filter trade-filter-instrument form-control"
        label="Instrument"
        value={filters.instrument}
        onChange={e => updateFilter('instrument', e.target.value)}
      />
      <TextInput
        type="text"
        className="trade-filter trade-filter-strategy form-control"
        label="Strategy"
        value={filters.strategy}
        onChange={e => updateFilter('strategy', e.target.value)}
      />
      <SelectInput
        className="trade-filter trade-filter-kind"
        value={filters.kind}
        choices={kindChoices}
        label="Kind"
        onChange={e => updateFilter('kind', e.target.value)}
      />

      <SelectInput
        className="trade-filter trade-filter-rating"
        value={filters.rating}
        choices={ratingChoices}
        label="Rating"
        onChange={e => updateFilter('rating', e.target.value)}
      />
      <div className="trade-filter trade-quick-create-date form-control">
        <ReactDatePicker
          className="trade-filter-entryDate"
          selected={filters.entryDate}
          onChange={d => updateFilter('entryDate', d)}
          placeholderText="Entry Date"
          dateFormat="MMMM d, yyyy"
          isClearable
        />
      </div>
      <div className="trade-filter trade-quick-create-date form-control">
        <ReactDatePicker
          className="trade-filter-exitDate"
          selected={filters.exitDate}
          onChange={d => updateFilter('exitDate', d)}
          placeholderText="Exit Date"
          dateFormat="MMMM d, yyyy"
          isClearable
        />
      </div>

      <div className="trade-filter">
        <ReactTags
          tags={tags.tags.tags}
          suggestions={[]}
          handleDelete={(i: number) => handleTagDelete('tags', i)}
          handleAddition={(tag: Tag) => handleTagAddition('tags', tag)}
          placeholder="Add Tag"
          allowNew
        />
      </div>
      <div className="trade-filter">
        <ReactTags
          tags={tags.emotions.tags}
          suggestions={[]}
          handleDelete={(i: number) => handleTagDelete('emotions', i)}
          handleAddition={(tag: Tag) => handleTagAddition('emotions', tag)}
          placeholder="Add Emotion"
          allowNew
        />
      </div>
      <div className="trade-filter-checkboxes grid-x grid-margin-x grid-margin-y">
        <div className="trade-filter form-check cell medium-6">
          <TristateCheckbox
            id="profit-tristate"
            className="form-check-input trade-filter-profit"
            onClick={val => updateFilter('profit', val)}
          />
          <label className="form-check-label" htmlFor="profit-tristate">
            Profitable
          </label>
        </div>
        <div className="trade-filter form-check cell medium-6">
          <TristateCheckbox
            id="hit-take-profit-tristate"
            className="form-check-input trade-filter-hit-take-profit"
            onClick={val => updateFilter('hitTakeProfit', val)}
          />
          <label
            className="form-check-label"
            htmlFor="hit-take-profit-tristate"
          >
            Hit Target
          </label>
        </div>
        <div className="trade-filter form-check cell medium-6">
          <TristateCheckbox
            id="flagged-tristate"
            className="form-check-input trade-filter-flagged"
            onClick={val => updateFilter('flagged', val)}
          />
          <label className="form-check-label" htmlFor="flagged-tristate">
            Flagged
          </label>
        </div>
        <div className="trade-filter form-check cell medium-6">
          <TristateCheckbox
            id="managed-tristate"
            className="form-check-input trade-filter-managed"
            onClick={val => updateFilter('managed', val)}
          />
          <label className="form-check-label" htmlFor="mangaged-tristate">
            Managed
          </label>
        </div>
      </div>
    </div>
  )
}

const mapDispatchToProps = (dispatch: Dispatch<Types.RootAction>) => ({
  // @ts-ignore
  onSetTradeFilters: (filters: Types.TradeFilters) =>
    dispatch(setTradeFilters(filters)),
})

export const TradeFilter = connect(
  null,
  mapDispatchToProps
)(TradeFilterComponent)
