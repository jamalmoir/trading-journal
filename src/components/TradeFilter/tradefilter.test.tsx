import * as React from 'react';
import { TradeFilterComponent } from './TradeFilter';
import { shallow } from 'enzyme';
import Types from 'Types';


describe("<TradeFilter />", () => {
  it('matches snapshot', () => {
		const onSetTradeFilters = jest.fn();
    const component = shallow(<TradeFilterComponent onSetTradeFilters={ onSetTradeFilters } />);
    expect(component).toMatchSnapshot();
	});

  it('calls onSetTradeFilters with the right arguments', () => {
		const onSetTradeFilters = jest.fn();
		const component = shallow(<TradeFilterComponent onSetTradeFilters={ onSetTradeFilters } />);

		const instrument = 'GBPJPY';
		const strategy = 'Cypher';
		const kind = 'long';
		const rating = 1;
		const entryDate: Types.TradeFilters['entryDate'] = null;
		const exitDate: Types.TradeFilters['exitDate'] = null;
		const profit: Types.TradeFilters['profit'] = null;
		const hitTakeProfit: Types.TradeFilters['hitTakeProfit'] = null;
		const flagged: Types.TradeFilters['flagged'] = null;
		const managed: Types.TradeFilters['managed'] = null;
		const tags: Types.TradeFilters['tags'] = null;
		const emotions: Types.TradeFilters['emotions'] = null;

		const expectedFilters: Types.TradeFilters = {
			instrument: instrument,
			strategy: strategy,
			kind: kind,
			rating: rating,
			entryDate: entryDate,
			exitDate: exitDate,
			profit: profit,
			hitTakeProfit: hitTakeProfit,
			flagged: flagged,
			managed: managed,
			tags: tags,
			emotions: emotions,
		};

		component.find('.trade-filter-instrument').simulate('change', { target: { value: instrument } });
		component.find('.trade-filter-strategy').simulate('change', { target: { value: strategy } });
		component.find('.trade-filter-kind').simulate('change', { target: { value: kind } });
		component.find('.trade-filter-rating').simulate('change', { target: { value: rating } });
		// component.find('.trade-filter-entry-date').simulate('change', { target: { value: entryDate } });
		// component.find('.trade-filter-exit-date').simulate('change', { target: { value: exitDate } });
		// component.find('.trade-filter-profit').simulate('click');
		// component.find('.trade-filter-hit-take-profit').simulate('click');
		// component.find('.trade-filter-flagged').simulate('click');
		// component.find('.trade-filter-flagged').simulate('click');
		// component.find('.trade-filter-managed').simulate('click');
		// component.find('.trade-filter-managed').simulate('click');
		// component.find('.trade-filter-tags').simulate('change', { target: { value: tags } });
		// component.find('.trade-filter-emotions').simulate('change', { target: { value: emotions } });

		expect(onSetTradeFilters).toBeCalledTimes(4);
		expect(JSON.stringify(onSetTradeFilters.mock.calls[3][0]) === JSON.stringify(expectedFilters)).toBeTruthy();
  });
});