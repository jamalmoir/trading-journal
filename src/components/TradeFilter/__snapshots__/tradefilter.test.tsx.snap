// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TradeFilter /> matches snapshot 1`] = `
<div
  className="trade-filter-panel row bg-light"
>
  <div
    className="col-sm-12"
  >
    Filters
  </div>
  <div
    className="trade-filter-group col-sm-3"
  >
    <Component
      className="trade-filter trade-filter-instrument form-control"
      onChange={[Function]}
      placeholder="Instrument"
      type="text"
      value={null}
    />
    <Component
      className="trade-filter trade-filter-strategy form-control"
      onChange={[Function]}
      placeholder="Strategy"
      type="text"
      value={null}
    />
    <select
      className="trade-filter trade-filter-kind custom-select"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Kind
      </option>
      <option
        value="long"
      >
        Long
      </option>
      <option
        value="short"
      >
        Short
      </option>
    </select>
  </div>
  <div
    className="trade-filter-group col-sm-3"
  >
    <select
      className="trade-filter trade-filter-rating custom-select"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      >
        Rating
      </option>
      <option
        value="-1"
      >
        -1
      </option>
      <option
        value="0"
      >
        0
      </option>
      <option
        value="1"
      >
        1
      </option>
    </select>
    <div
      className="trade-filter trade-quick-create-date form-control"
    >
      <DatePicker
        allowSameDay={false}
        className="trade-filter-entryDate"
        dateFormat="MMMM d, yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        isClearable={true}
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Entry Date"
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={null}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </div>
    <div
      className="trade-filter trade-quick-create-date form-control"
    >
      <DatePicker
        allowSameDay={false}
        className="trade-filter-exitDate"
        dateFormat="MMMM d, yyyy"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        inlineFocusSelectedMonth={false}
        isClearable={true}
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Exit Date"
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        selected={null}
        shouldCloseOnSelect={true}
        showMonthYearPicker={false}
        showTimeInput={false}
        showTimeSelect={false}
        strictParsing={false}
        timeCaption="Time"
        timeInputLabel="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </div>
  </div>
  <div
    className="trade-filter-group col-sm-4"
  >
    <div
      className="trade-filter"
    >
      <Component
        allowNew={true}
        handleAddition={[Function]}
        handleDelete={[Function]}
        placeholder="Add Tag"
        suggestions={Array []}
        tags={Array []}
      />
    </div>
    <div
      className="trade-filter"
    >
      <Component
        allowNew={true}
        handleAddition={[Function]}
        handleDelete={[Function]}
        placeholder="Add Emotion"
        suggestions={Array []}
        tags={Array []}
      />
    </div>
  </div>
  <div
    className="trade-filter-group col-sm-2"
  >
    <div
      className="trade-filter form-check"
    >
      <Component
        className="form-check-input trade-filter-profit"
        id="profit-tristate"
        onClick={[Function]}
      />
      <label
        className="form-check-label"
        htmlFor="profit-tristate"
      >
        Profitable
      </label>
    </div>
    <div
      className="trade-filter form-check"
    >
      <Component
        className="form-check-input trade-filter-hit-take-profit"
        id="hit-take-profit-tristate"
        onClick={[Function]}
      />
      <label
        className="form-check-label"
        htmlFor="hit-take-profit-tristate"
      >
        Hit Target
      </label>
    </div>
    <div
      className="trade-filter form-check"
    >
      <Component
        className="form-check-input trade-filter-flagged"
        id="flagged-tristate"
        onClick={[Function]}
      />
      <label
        className="form-check-label"
        htmlFor="flagged-tristate"
      >
        Flagged
      </label>
    </div>
    <div
      className="trade-filter form-check"
    >
      <Component
        className="form-check-input trade-filter-managed"
        id="managed-tristate"
        onClick={[Function]}
      />
      <label
        className="form-check-label"
        htmlFor="mangaged-tristate"
      >
        Managed
      </label>
    </div>
  </div>
</div>
`;
